ucsb-cs156-s24/proj-happycows-s24-4pm-6
# Issues B3a & B3b: Add code that monitors role changes and gatekeepers suspended users at login time

## Overview

In this PR, I address issues B3a and B3b together as my initial attempt to integrate them separately proved much more tedious, with testing, than by doing them together. I add code that monitors role changes, similar to gaucho ride and organic. With this code that constantly checks if a user is suspended, I update their login privileges and redirect them to an error page if they are suspended. A message is also sent to indicate they have been suspended. If the admin would like to restore a user, they can do so and the user will be redirected to the login page upon refreshing.

Dokku deployment: https://happycows-nbelle1-dev.dokku-06.cs.ucsb.edu

## Screenshots: Below are the screens you should see when following the steps I outline below:
<img width="1186" alt="Screenshot 2024-05-24 at 1 01 46 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/8aedf381-71a3-4df7-9e77-a9483e76f362">
<img width="1192" alt="Screenshot 2024-05-24 at 12 55 23 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/cc725eb9-5555-4b4b-8881-e86b72d92796">
<img width="1195" alt="Screenshot 2024-05-24 at 12 55 29 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/2e6d5fe1-8b92-4383-b179-636c7e9413f9">
<img width="1167" alt="Screenshot 2024-05-24 at 12 55 37 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/fc465325-81e2-43e5-839a-29d72f0dde12">
<img width="1184" alt="Screenshot 2024-05-24 at 12 55 46 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/d64e99b8-273e-4aee-9616-559bdbfc3e58">
<img width="1177" alt="Screenshot 2024-05-24 at 12 55 57 PM" src="https://github.com/ucsb-cs156-s24/proj-happycows-s24-4pm-6/assets/89555340/d6906c6a-a2bc-4cd4-bf93-c81d0e3b604c">


## Future Possibilities
Some possible points could be:
- This is part of EPIC: Admin can suspend/restore users


## Validation

1. Download this branch.
2. Run the project in two windows: as an admin and as a user.
3. Go to swagger UI on the admin window and find "User Information"
4. Test between suspending and restoring the user by ID.
5. See your privileges change in other window as the user.

## Tests


## Linked Issues

Closes #40 
Closes #46
